services/platform/platform.go

```diff
@@ -0,0 +1,200 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package platform provides authentication/authorization functionality.
+package platform
+
+import (
+	"context"
+	"net"
+	"os"
+	"time"
+
+	api "github.com/percona-platform/saas/gen/auth/external"
+	"github.com/percona/pmm/utils/tlsconfig"
+	"github.com/percona/pmm/version"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm-managed/models"
+)
+
+const (
+	defaultHost            = "check.percona.com:443"
+	defaultRefreshInterval = 24 * time.Hour
+	timeout                = 5 * time.Second
+
+	envHost            = "PERCONA_TEST_AUTH_HOST"
+	envRefreshInterval = "PERCONA_TEST_REFRESH_INTERVAL"
+
+	authType = "PP-v1beta1" // TODO Change to PP-1 after auth API release
+
+)
+
+var ErrNoActiveSessions = errors.New("no active sessions") //nolint:golint
+
+// Service is responsible for interactions with Percona platform.
+type Service struct {
+	db              *reform.DB
+	host            string
+	refreshInterval time.Duration
+	l               *logrus.Entry
+}
+
+// New returns platform Service.
+func New(db *reform.DB) *Service {
+	l := logrus.WithField("component", "auth")
+
+	s := Service{
+		host:            defaultHost,
+		refreshInterval: defaultRefreshInterval,
+		db:              db,
+		l:               l,
+	}
+
+	if h := os.Getenv(envHost); h != "" {
+		l.Warnf("Host changed to %s.", h)
+		s.host = h
+	}
+
+	if d, err := time.ParseDuration(os.Getenv(envRefreshInterval)); err == nil && d > 0 {
+		l.Warnf("Refresh interval changed to %s.", d)
+		s.refreshInterval = d
+	}
+
+	return &s
+}
+
+// Run refreshes Percona Platform session every interval until context is canceled.
+func (s *Service) Run(ctx context.Context) {
+	ticker := time.NewTicker(s.refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+		// continue with next loop iteration
+		case <-ctx.Done():
+			return
+		}
+
+		err := s.RefreshSession(ctx)
+		if err != nil {
+			s.l.Warn(err)
+		}
+		if err != nil && err != ErrNoActiveSessions {
```

Comment:
Two checks for `err != nil`, so we can log the same error twice?

Reply:
Yep, it's a mistake. Removed.